api/v1alpha1/securitygroup_types.go

```diff
@@ -26,40 +26,88 @@ type RuleDirection string
+
 // +kubebuilder:validation:Enum:=IPv4;IPv6
 // +kubebuilder:validation:MinLength:=1
-// +kubebuilder:validation:MaxLength:=16
+// +kubebuilder:validation:MaxLength:=4
 type Ethertype string
 
+const (
+	EtherTypeIPv4 Ethertype = "IPv4"
+	EtherTypeIPv6 Ethertype = "IPv6"
+)
+
 // SecurityGroupRule defines a Security Group rule
 // +kubebuilder:validation:MinProperties:=1
+// +kubebuilder:validation:XValidation:rule="has(self.portRangeMin) == has(self.portRangeMax)",message="PortRangeMin and PortRangeMax should exist together"
+// +kubebuilder:validation:XValidation:rule="(!has(self.portRangeMin) || !has(self.portRangeMax)) || has(self.protocol)",message="Protocol must exist when PortRangeMin or PortRangeMax exist"
+// +kubebuilder:validation:XValidation:rule="(!has(self.portRangeMin) || !has(self.portRangeMax))|| (self.portRangeMin <= self.portRangeMax)",message="PortRangeMax should be equal or greater than PortRangeMin"
+// +kubebuilder:validation:XValidation:rule="!has(self.protocol) || !(self.protocol == 'icmp' || self.protocol == 'icmpv6') || (self.portRangeMin >= 0 && self.portRangeMin <= 255 && self.portRangeMax >= 0 && self.portRangeMax <= 255)",message="When protocol is ICMP or ICMPv6 PortRangeMin and PortRangeMax should be between 0 and 255"
+// +kubebuilder:validation:XValidation:rule="!has(self.remoteIPPrefix) || (isCIDR(self.remoteIPPrefix) && cidr(self.remoteIPPrefix).ip().family() == 4 && self.ethertype == 'IPv4') || (isCIDR(self.remoteIPPrefix) && cidr(self.remoteIPPrefix).ip().family() == 6 && self.ethertype == 'IPv6')",message="RemoteIPPrefix should be a valid CIDR and match the Ethertype"
 type SecurityGroupRule struct {
 	// Description of the existing resource
 	// +optional
+
 	Description *OpenStackDescription `json:"description,omitempty"`
 
 	// Direction represents the direction in which the security group rule
 	// is applied. Can be ingress or egress.
+	// +optional
 	Direction *RuleDirection `json:"direction,omitempty"`
 
 	// RemoteAddressGroupId (Not in gophercloud)
-
 	// RemoteGroupID
 	RemoteGroupID *UUID `json:"remoteGroupID,omitempty"`
 
-	// RemoteIPPrefix
+	// RemoteIPPrefix is an IP address block. Should match the Ethertype (IPv4 or IPv6)
+	// +optional
 	RemoteIPPrefix *CIDR `json:"remoteIPPrefix,omitempty"`
 
 	// Protocol is the IP protocol can be represented by a string, an
 	// integer, or null
+	// +optional
 	Protocol *Protocol `json:"protocol,omitempty"`
 
 	// EtherType must be IPv4 or IPv6, and addresses represented in CIDR
 	// must match the ingress or egress rules.
-	Ethertype *Ethertype `json:"ethertype,omitempty"`
-
-	PortRangeMin *int32 `json:"portRangeMin,omitempty"`
-	PortRangeMax *int32 `json:"portRangeMax,omitempty"`
+	// +kubebuilder:validation:Required
+	Ethertype *Ethertype `json:"ethertype"`
+	// If the protocol is [tcp, udp, dccp sctp,udplite] this value must be less than
+	// or equal to the PortRangeMax attribute value.
+	// If the protocol is ICMP, this value must be an ICMP type
```

Comment:
```If the protocol is ICMP, this value must be an ICMP type```

Before reading the code, I can't immediately find a meaning to this sentence

Reply:
When setting up ICMP, I would naïvely leave PortRangeMin and PortRangeMax empty. Would I be wrong?

Reply:
First I have a typo, PorRangeMin -> icmp type, PortRangeMax - icmp code
Can you suggest an alternative to the comment?
At least from my testing with the openstack client you can leave them out (Then I think the rule catch all the ICMP types)